README.rst
sprockets.mixins.cors
=====================
This library exposes a mix-in that adds some useful CORS_ hooks over
a basic ``tornado.web.RequestHandler``.

.. _CORS: http://www.w3.org/TR/cors/


docs/history.rst
Version History
===============

`1.0.1`_
--------
 - fix awaiting the returned coroutine if the handler's ``prepare`` function is async

`0.1.1`_
--------
 - minor packaging changes

`0.1.0`_
--------
 - initial implementation including:

   - :class:`sprockets.mixins.cors.CORSMixin`
   - :class:`sprockets.mixins.cors.CORSSettings`
   - Support for the following pre-flight request headers:

     - :mailheader:`Origin`
     - :mailheader:`Access-Control-Request-Method`
     - :mailheader:`Access-Control-Request-Headers`

   - Support for the following pre-flight response headers:

     - :mailheader:`Access-Control-Allow-Origin`
     - :mailheader:`Access-Control-Allow-Methods`
     - :mailheader:`Access-Control-Allow-Credentials`
     - :mailheader:`Access-Control-Allow-Headers`

   - Support for the following in-line response headers:

     - :mailheader:`Access-Control-Allow-Origin`
     - :mailheader:`Access-Control-Allow-Credentials`

.. _1.0.1: https://github.com/sprockets/sprockets.mixins.cors/compare/1.0.0...1.0.1
.. _0.1.1: https://github.com/sprockets/sprockets.mixins.cors/compare/0.1.0...0.1.1
.. _0.1.0: https://github.com/sprockets/sprockets.mixins.cors/compare/0.0.0...0.1.0


tests.py
from __future__ import absolute_import

import unittest

from tornado import testing, web

from examples import SimpleRequestHandler
from sprockets.mixins import cors


class OptionSupportTests(testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application([web.url('/', SimpleRequestHandler)])

    def test_that_options_is_supported(self):
        response = self.fetch('/', method='OPTIONS')
        self.assertIn(response.code, range(200, 300))

    def test_that_options_includes_supported_methods(self):
        response = self.fetch('/', method='OPTIONS')
        self.assertEqual(response.headers['Allow'],
                         ','.join(SimpleRequestHandler.SUPPORTED_METHODS))


class PreflightTests(testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application(
            [web.url('/', SimpleRequestHandler),
             web.url('/private', SimpleRequestHandler, {'creds': True}),
             web.url('/reqheader', SimpleRequestHandler,
                     {'req_headers': ['Correlation-ID']})],
            cors_origins=['http://host.example.com'],
        )

    def test_that_options_understands_origin_header(self):
        response = self.fetch('/', method='OPTIONS',
                              headers={'Origin': 'http://host.example.com',
                                       'Access-Control-Request-Method': 'GET'})
        self.assertIn(response.code, range(200, 300))
        self.assertEqual(response.headers['Access-Control-Allow-Origin'],
                         'http://host.example.com')

    def test_that_preflight_fails_for_unacceptable_origin(self):
        response = self.fetch('/', method='OPTIONS',
                              headers={'Origin': 'https://host.example.com',
                                       'Access-Control-Request-Method': 'GET'})
        self.assertEqual(response.code, 403)

    def test_that_preflight_fails_for_unacceptable_method(self):
        response = self.fetch(
            '/', method='OPTIONS',
            headers={'Origin': 'http://host.example.com',
                     'Access-Control-Request-Method': 'POST'})
        self.assertEqual(response.code, 403)

    def test_that_preflight_fails_when_missing_request_method(self):
        response = self.fetch('/', method='OPTIONS',
                              headers={'Origin': 'http://host.example.com'})
        self.assertEqual(response.code, 403)

    def test_that_preflight_generates_allow_credentials_appropriately(self):
        response = self.fetch('/private', method='OPTIONS',
                              headers={'Origin': 'http://host.example.com',
                                       'Access-Control-Request-Method': 'GET'})
        self.assertEqual(response.headers['Access-Control-Allow-Credentials'],
                         'true')

    def test_that_preflight_fails_for_unsupported_request_header(self):
        response = self.fetch('/', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': 'Correlation-ID',
        })
        self.assertEqual(response.code, 403)

    def test_that_preflight_succeeds_for_simple_request_headers(self):
        request_headers = ', '.join(s.title()
                                    for s in cors.SIMPLE_REQUEST_HEADERS)
        response = self.fetch('/', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': request_headers,
        })
        self.assertIn(response.code, range(200, 300))

    def test_that_preflight_does_not_advertise_simple_request_headers(self):
        request_headers = ', '.join(s.title()
                                    for s in cors.SIMPLE_REQUEST_HEADERS)
        response = self.fetch('/', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': request_headers,
        })
        self.assertIn(response.code, range(200, 300))
        self.assertNotIn('Access-Control-Allow-Headers', response.headers)

    def test_that_preflight_does_not_advertise_simple_request_headers(self):
        request_headers = ', '.join(s.title()
                                    for s in cors.SIMPLE_REQUEST_HEADERS)
        response = self.fetch('/', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': request_headers,
        })
        self.assertIn(response.code, range(200, 300))
        self.assertNotIn('Access-Control-Allow-Headers', response.headers)

    def test_that_preflight_advertises_custom_request_headers(self):
        response = self.fetch('/reqheader', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': 'Correlation-ID',
        })
        self.assertIn(response.code, range(200, 300))
        self.assertIn('correlation-id',
                      response.headers['Access-Control-Allow-Headers'].lower())


class StandardRequestTests(testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application(
            [web.url('/', SimpleRequestHandler),
             web.url('/private', SimpleRequestHandler, {'creds': True})],
            cors_origins=['http://host.example.com'],
        )

    def test_that_get_response_includes_allow_origin(self):
        response = self.fetch('/', headers={
            'Origin': 'http://host.example.com'})
        self.assertEqual(response.headers['Access-Control-Allow-Origin'],
                         'http://host.example.com')

    def test_that_get_response_includes_allow_creds_when_secure(self):
        response = self.fetch('/private', headers={
            'Origin': 'http://host.example.com'})
        self.assertEqual(response.headers['Access-Control-Allow-Origin'],
                         'http://host.example.com')
        self.assertEqual(response.headers['Access-Control-Allow-Credentials'],
                         'true')

    def test_that_get_skips_allow_origin_on_missing_origin_header(self):
        response = self.fetch('/')
        self.assertNotIn('Access-Control-Allow-Origin', response.headers)

    def test_that_get_skips_allow_origin_when_handler_finishes(self):
        response = self.fetch('/', headers={
            'Origin': 'http://host.example.com', 'X-Fail': 'yes please'})
        self.assertNotIn('Access-Control-Allow-Origin', response.headers)


tests/tests.py
from __future__ import absolute_import

import asyncio

from examples import SimpleRequestHandler
from tornado import testing, web

from sprockets.mixins import cors


class OptionSupportTests(testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application([web.url('/', SimpleRequestHandler)])

    def test_that_options_is_supported(self):
        response = self.fetch('/', method='OPTIONS')
        self.assertIn(response.code, range(200, 300))

    def test_that_options_includes_supported_methods(self):
        response = self.fetch('/', method='OPTIONS')
        self.assertEqual(response.headers['Allow'],
                         ','.join(SimpleRequestHandler.SUPPORTED_METHODS))


class PreflightTests(testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application(
            [web.url('/', SimpleRequestHandler),
             web.url('/private', SimpleRequestHandler, {'creds': True}),
             web.url('/reqheader', SimpleRequestHandler,
                     {'req_headers': ['Correlation-ID']})],
            cors_origins=['http://host.example.com'],
        )

    def test_that_options_understands_origin_header(self):
        response = self.fetch('/', method='OPTIONS',
                              headers={'Origin': 'http://host.example.com',
                                       'Access-Control-Request-Method': 'GET'})
        self.assertIn(response.code, range(200, 300))
        self.assertEqual(response.headers['Access-Control-Allow-Origin'],
                         'http://host.example.com')

    def test_that_preflight_fails_for_unacceptable_origin(self):
        response = self.fetch('/', method='OPTIONS',
                              headers={'Origin': 'https://host.example.com',
                                       'Access-Control-Request-Method': 'GET'})
        self.assertEqual(response.code, 403)

    def test_that_preflight_fails_for_unacceptable_method(self):
        response = self.fetch(
            '/', method='OPTIONS',
            headers={'Origin': 'http://host.example.com',
                     'Access-Control-Request-Method': 'POST'})
        self.assertEqual(response.code, 403)

    def test_that_preflight_fails_when_missing_request_method(self):
        response = self.fetch('/', method='OPTIONS',
                              headers={'Origin': 'http://host.example.com'})
        self.assertEqual(response.code, 403)

    def test_that_preflight_generates_allow_credentials_appropriately(self):
        response = self.fetch('/private', method='OPTIONS',
                              headers={'Origin': 'http://host.example.com',
                                       'Access-Control-Request-Method': 'GET'})
        self.assertEqual(response.headers['Access-Control-Allow-Credentials'],
                         'true')

    def test_that_preflight_fails_for_unsupported_request_header(self):
        response = self.fetch('/', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': 'Correlation-ID',
        })
        self.assertEqual(response.code, 403)

    def test_that_preflight_succeeds_for_simple_request_headers(self):
        request_headers = ', '.join(s.title()
                                    for s in cors.SIMPLE_REQUEST_HEADERS)
        response = self.fetch('/', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': request_headers,
        })
        self.assertIn(response.code, range(200, 300))

    def test_that_preflight_does_not_advertise_simple_request_headers(self):
        request_headers = ', '.join(s.title()
                                    for s in cors.SIMPLE_REQUEST_HEADERS)
        response = self.fetch('/', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': request_headers,
        })
        self.assertIn(response.code, range(200, 300))
        self.assertNotIn('Access-Control-Allow-Headers', response.headers)

    def test_that_preflight_advertises_custom_request_headers(self):
        response = self.fetch('/reqheader', method='OPTIONS', headers={
            'Origin': 'http://host.example.com',
            'Access-Control-Request-Method': 'GET',
            'Access-Control-Request-Headers': 'Correlation-ID',
        })
        self.assertIn(response.code, range(200, 300))
        self.assertIn('correlation-id',
                      response.headers['Access-Control-Allow-Headers'].lower())


class StandardRequestTests(testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application(
            [web.url('/', SimpleRequestHandler),
             web.url('/private', SimpleRequestHandler, {'creds': True})],
            cors_origins=['http://host.example.com'],
        )

    def test_that_get_response_includes_allow_origin(self):
        response = self.fetch('/', headers={
            'Origin': 'http://host.example.com'})
        self.assertEqual(response.headers['Access-Control-Allow-Origin'],
                         'http://host.example.com')

    def test_that_get_response_includes_allow_creds_when_secure(self):
        response = self.fetch('/private', headers={
            'Origin': 'http://host.example.com'})
        self.assertEqual(response.headers['Access-Control-Allow-Origin'],
                         'http://host.example.com')
        self.assertEqual(response.headers['Access-Control-Allow-Credentials'],
                         'true')

    def test_that_get_skips_allow_origin_on_missing_origin_header(self):
        response = self.fetch('/')
        self.assertNotIn('Access-Control-Allow-Origin', response.headers)

    def test_that_get_skips_allow_origin_when_handler_finishes(self):
        response = self.fetch('/', headers={
            'Origin': 'http://host.example.com', 'X-Fail': 'yes please'})
        self.assertNotIn('Access-Control-Allow-Origin', response.headers)


class PrepareCoroutineTests(testing.AsyncHTTPTestCase):
    def setUp(self):
        super().setUp()
        self.coroutine_has_finished = False

    def make_async_handler_class(self):
        test_instance = self

        class AsyncMixin:
            @staticmethod
            async def prepare():
                await asyncio.sleep(0)
                test_instance.coroutine_has_finished = True

        class AsyncPrepareHandler(cors.CORSMixin, AsyncMixin,
                                  web.RequestHandler):
            def get(self):
                self.set_status(204)
                self.finish()

        return AsyncPrepareHandler

    def get_app(self):
        async_prepare_handler = self.make_async_handler_class()
        return web.Application([web.url('/', async_prepare_handler)])

    def test_prepare_coroutine_is_awaited(self):
        self.fetch('/')
        self.assertTrue(self.coroutine_has_finished)


examples.py
#!/usr/bin/env python

import logging

from tornado import ioloop, web

from sprockets.mixins import cors


class SimpleRequestHandler(cors.CORSMixin, web.RequestHandler):
    """Very simple request handler that CORS enables the GET endpoint."""

    def initialize(self, creds=False, req_headers=None):
        super(SimpleRequestHandler, self).initialize()
        self.cors.allowed_methods.add('GET')
        self.cors.credentials_supported = creds
        if req_headers:
            self.cors.request_headers.update(hdr.lower()
                                             for hdr in req_headers)

    def prepare(self):
        # This is used to test that the mixin does not interfere
        # with request failures.  You really shouldn't call super()
        # after you explicitly finish() but anyway...
        if 'X-Fail' in self.request.headers:
            self.set_status(400)
            self.finish()

        return super(SimpleRequestHandler, self).prepare()

    def get(self):
        self.set_status(204)
        self.finish()


if __name__ == '__main__':
    logging.basicConfig(
        level=logging.DEBUG,
        format='%(levelname)-8s %(name)s: %(message)s')
    app = web.Application([
        web.url('/public', SimpleRequestHandler),
        web.url('/private', SimpleRequestHandler, {'creds': True}),
    ], cors_origins=['http://www.example.com'], debug=True)
    app.listen(8000)

    iol = ioloop.IOLoop.instance()
    try:
        iol.start()
    except KeyboardInterrupt:
        logging.info('stopping IOLoop')
        iol.add_callback(iol.stop)


sprockets/mixins/cors.py
"""

Tornado RequestHandler mix-in for implementing a CORS enabled endpoint.

The CORS_ specification describes a method of securing javascript access
to web resources across access domains.  This module implements a mix-in
to be used with :class:`tornado.web.RequestHandler` that provides much
of the functionality required by CORS_.

.. _CORS: http://www.w3.org/TR/cors/

"""
import asyncio

version_info = (1, 0, 1)
__version__ = '.'.join(str(v) for v in version_info)

SIMPLE_REQUEST_HEADERS = frozenset(('accept', 'accept-language',
                                    'content-language'))
"""
Request headers that get special treatment per CORS.

These are considered "simple headers" and are always acceptable to
send in CORS-enabled requests.

"""


class CORSSettings(object):
    """
    Configures the CORS behavior.

    .. attribute::allowed_methods

       The :class:`set` of CORS accepted HTTP methods.  This controls
       the :mailheader:`Access-Control-Allow-Methods` response header.

    .. attribute::allowed_origins

       The :class:`set` of origins that are allowed for the endpoint.
       This controls the :mailheader:`Access-Control-Allow-Origin`
       response header.  If the requested origin is in this set, then
       the origin is allowed; otherwise, a :http:statuscode:`403` is returned.

    .. attribute::credentials_supported

       Should the mix-in generate the
       :mailheader:`Access-Control-Allow-Credentials` header in the
       response.

    .. attribute::request_headers

       A :class:`set` of header names that are acceptable in cross-origin
       requests.  Headers added to this set **MUST** be lower-cased before
       adding them to the set.

    """
    def __init__(self):
        self.allowed_methods = set()
        self.allowed_origins = set()
        self.credentials_supported = False
        self.request_headers = {header.lower()
                                for header in SIMPLE_REQUEST_HEADERS}


class CORSMixin(object):
    """
    Mix this in over a :class:`tornado.web.RequestHandler` for CORS support.

    .. attribute::cors

       A :class:`.CORSSettings` instance that controls the behavior of
       the mix-in.

    """

    def initialize(self, **kwargs):
        self.cors = CORSSettings()
        self.cors.allowed_origins.update(self.settings.get('cors_origins', []))
        super(CORSMixin, self).initialize(**kwargs)

    async def prepare(self):
        maybe_coro = super(CORSMixin, self).prepare()
        if asyncio.iscoroutine(maybe_coro):
            await maybe_coro

        if not self._finished and self.request.method != 'OPTIONS':
            origin = self.request.headers.get('Origin')
            if origin in self.cors.allowed_origins:
                self.set_header('Access-Control-Allow-Origin', origin)
                if self.cors.credentials_supported:
                    self.set_header('Access-Control-Allow-Credentials', 'true')

    def options(self):
        """
        Respond to an :http:method:OPTIONS request.

        This method relies on :attr:`self.SUPPORTED_METHODS` for the
        content of the :http:header:Allow response header.  The CORS
        specific headers are generated based on the :attr:`.cors`
        attribute.

        """
        self.set_header('Allow', ','.join(self.SUPPORTED_METHODS))
        self.set_status(204)
        if 'Origin' in self.request.headers:
            if self._cors_preflight_checks():
                self._build_preflight_response(self.request.headers['Origin'])
            else:
                self.set_status(403)

    def _cors_preflight_checks(self):
        try:
            origin = self.request.headers['Origin']
            method = self.request.headers['Access-Control-Request-Method']
            headers = self.request.headers.get(
                'Access-Control-Request-Headers', '')
        except KeyError:
            return False

        headers = _filter_headers(headers, self.cors.request_headers)
        return (origin in self.cors.allowed_origins and
                method in self.cors.allowed_methods and
                len(headers) == 0)

    def _build_preflight_response(self, origin):
        self.set_header('Access-Control-Allow-Origin', origin)
        self.set_header('Access-Control-Allow-Methods',
                        ','.join(self.cors.allowed_methods))
        if self.cors.credentials_supported:
            self.set_header('Access-Control-Allow-Credentials', 'true')
        exposed_headers = self.cors.request_headers - SIMPLE_REQUEST_HEADERS
        if exposed_headers:
            self.set_header('Access-Control-Allow-Headers',
                            ','.join(exposed_headers))

    def _clear_headers_for_304(self) -> None:
        # Overrides '_clear_headers_for_304' method from
        # web.RequestHandler to not clear the ALLOW header when the status
        # code is set to 204. This is bug in Tornado, which is fixed in the
        # Tornado v6.1 but not yet released. This method can be removed once
        # it is updated to Tornado v6.1
        headers = [
            'Content-Encoding',
            'Content-Language',
            'Content-Length',
            'Content-MD5',
            'Content-Range',
            'Content-Type',
            'Last-Modified',
        ]
        for h in headers:
            self.clear_header(h)


def _filter_headers(header_str, simple_headers):
    header_str = header_str.lower().replace(' ', '').replace('\t', '')
    if not header_str:
        return set()

    header_set = {str(value) for value in header_str.split(',')}
    header_set.difference_update(simple_headers)
    header_set.difference_update('')
    return header_set
